src/RemediationRoleSaver.cpp

```diff
@@ -79,35 +151,195 @@ void RemediationSaverBase::saveToFile(const QString& filename)
+void ResultBasedLibraryRemediationSaver::saveToFile(const QString& filename)
 {
-    mSaveMessage = QObject::tr("Save remediation role as a puppet manifest");
-    mFiletypeExtension = "pp";
-    mFiletypeTemplate = QObject::tr("puppet manifest (*.%1)");
-    mFixTemplate = QObject::tr("urn:xccdf:fix:script:puppet");
+    struct oscap_source* source = oscap_source_new_from_file(mArfFile.fileName().toUtf8().constData());
+    oscap_document_type_t document_type = oscap_source_get_scap_type(source);
+    if (document_type != OSCAP_DOCUMENT_ARF) {
+        throw std::runtime_error("Expected an ARF file");
+    }
+
+    struct ds_rds_session* arf_session = ds_rds_session_new_from_source(source);
+    if (arf_session == NULL) {
+        throw std::runtime_error("Couldn't open ARF session");
+    }
+    struct oscap_source *report_source = ds_rds_session_select_report(arf_session, NULL);
+    if (report_source == NULL) {
+        throw std::runtime_error("Couldn't get report source from the ARF session");
+    }
+    struct oscap_source *report_request_source = ds_rds_session_select_report_request(arf_session, NULL);
+    if (report_request_source == NULL) {
+        throw std::runtime_error("Couldn't get report request source from the ARF session");
+    }
+
+    struct xccdf_session* session = xccdf_session_new_from_source(oscap_source_clone(report_request_source));
+    if (xccdf_session_add_report_from_source(session, oscap_source_clone(report_source))) {
+        throw std::runtime_error("Couldn't get report request source from the ARF session");
+    }
+    oscap_source_free(source);
+
+    if (session == NULL)
+        throw std::runtime_error("Couldn't get XCCDF session from the report source");
+
+    xccdf_session_set_loading_flags(session, XCCDF_SESSION_LOAD_XCCDF);
+    if (xccdf_session_load(session) != 0)
+        throw std::runtime_error("Couldn't load XCCDF");
+
+    // "" for profile should work, as it works when passed to the command-line oscap
+    if (xccdf_session_build_policy_from_testresult(session, "") != 0)
+        throw std::runtime_error("Couldn't get build policy from testresult");
+
+    struct xccdf_policy* policy = xccdf_session_get_xccdf_policy(session);
+    struct xccdf_result* result = xccdf_policy_get_result_by_id(policy, xccdf_session_get_result_id(session));
+    /* Result-oriented fixes */
+
+    QFile outputFile(filename);
+    outputFile.open(QIODevice::WriteOnly);
+
+    const int rc = xccdf_policy_generate_fix(policy, result, mTemplateString.toUtf8().constData(), outputFile.handle());
+    if (!outputFile.flush())
+    {
+        throw std::runtime_error("Could not write to the destination location.");
+    }
+    outputFile.close();
+    ds_rds_session_free(arf_session);
+    xccdf_session_free(session);
+
+    if (rc != 0)
+    {
+        const char* err = oscap_err_desc();
+        if (err == NULL)
+            err = "Unknown error";
+        throw std::runtime_error(err);
```

Comment:
what happens if this throws an exception? It seems to me that the whole app would crash.

Reply:
No, it is caught up in `selectFilenameAndSaveRole` like all other exceptions. I have verified it and it works.